### ISSUE TITLE
Segmentation Fault when using SSL to connect inside a Debian container (node:18)

### ISSUE THREAD

**Issue by samfweb** [prisma/prisma#13478]
### Bug description

Prisma causes the process to crash when trying to connect to a hosted database that requires SSL when run inside a Debian container (node:18).

(note the log in the image below saying connecting to localhost is a leftover from a previous run...local connections without SSL work fine and the seg fault was caused connecting to a cloud-hosted db)
![image](https://user-images.githubusercontent.com/44712417/170181125-c597f7bd-f344-4d91-a87c-56d9fe8a0ee0.png)

This doesn't occur on windows machines and I've verified the connection and ssl parameters are correct by using node-pg to connect to the database without problems. 

I've provided the node-pg client connecting successfully alongside the segfault in the repo below.

### How to reproduce

Steps in README.md in example repo: 
https://github.com/samfweb/prisma-tsdb-ssl-segfault 

The repo uses TimescaleDB (a postgres extension).

### Expected behavior

Prisma should connect successfully and not crash the process.

### Prisma information

N/A

### Environment & setup

- Database: PostgreSQL (with TimescaleDB extension)
- Node.js version: v18.2.0
- OS: node:18 docker image (uses Debian)
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye

### Prisma Version

```
prisma                  : 3.14.0
@prisma/client          : 3.14.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at ../../../root/.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at ../../../root/.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at ../../../root/.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt 2b0c12756921c891fec4f68d9444e18c7d5d4a6a (at ../../../root/.npm/_npx/2778af9cee32ff87/node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : 2b0c12756921c891fec4f68d9444e18c7d5d4a6a
Studio                  : 0.460.0
Preview Features        : interactiveTransactions
```

**Comment by pantharshit00** [comment]
Can you try again once with Node 16?

**Comment by samfweb** [comment]
Node 16 works! Thanks for the suggestion.

**Comment by p30arena** [comment]
@pantharshit00 

I was using node v18 and faced this error too, then switched to v16 and the error is gone.

**Comment by laozhu** [comment]
I am using nodejs-18 on Debian, facing the same problem here, any workaround for this problem?

# uname -a
```
Linux xxoo.cloud 5.10.0-14-amd64 #1 SMP Debian 5.10.113-1 (2022-04-29) x86_64 GNU/Linux
```

# node -v
```
v18.5.0
```

# prisma -v
```
Environment variables loaded from .env
prisma                  : 4.0.0
@prisma/client          : 4.0.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine da41d2bb3406da22087b849f0e911199ba4fbf11 (at ../../usr/lib/node_modules/prisma/node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli da41d2bb3406da22087b849f0e911199ba4fbf11 (at ../../usr/lib/node_modules/prisma/node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core da41d2bb3406da22087b849f0e911199ba4fbf11 (at ../../usr/lib/node_modules/prisma/node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt da41d2bb3406da22087b849f0e911199ba4fbf11 (at ../../usr/lib/node_modules/prisma/node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : da41d2bb3406da22087b849f0e911199ba4fbf11
Studio                  : 0.465.0
```

**Comment by aqrln** [comment]
https://github.com/prisma/prisma/issues/10649#issuecomment-1247961614

**Comment by aqrln** [comment]
Thank you for the report!

I'm closing this issue in favour of https://github.com/prisma/prisma/issues/10649 as a duplicate to consolidate them and make it easier to keep track of the issue. Please post any questions or additional details in that thread, or open a new issue if you experience a similar problem under completely different conditions (e.g., with other DBMS or operating systems).

While we are looking for the best way to fix or prevent the underlying problem, the recommended workaround for affected users is to make sure that the versions of OpenSSL used by Node.js and the system OpenSSL match:
* either by downgrading Node.js to 16.x,
* or by upgrading the operating system (or Docker image) to use OpenSSL 3 — this might be less suitable for production systems, but this is a good option for those who only encounter the issue on their local machines during development.

Alternatively, if you can't do that and you need to stay with Node.js 18 on a system with OpenSSL 1.1.x, you can switch to the [binary engine](https://www.prisma.io/docs/concepts/components/prisma-engines/query-engine#defining-the-query-engine-type-for-prisma-client) in Prisma.

**Comment by Magofoco** [comment]
I use node version`v19.8.1`, I had the same error. I fixed as the following:

1) Added   engineType = "binary" in `schema.prisma` in the `client`
```
generator client {
  provider = "prisma-client-js"
  engineType = "binary"
}

```
2) Run `npx prisma migrate dev --name init`
3) Run `npx prisma generate`

**Comment by janpio** [comment]
Please don't use that as the solution, but only a temporary workaround. https://github.com/prisma/prisma/issues/10649 now has a development version that fixes the problem - please take a look at that as well and let us know if this works for you and properly _fixes_ the problem. Thanks.

**Comment by Magofoco** [comment]
> Please don't use that as the solution, but only a temporary workaround. #10649 now has a development version that fixes the problem - please take a look at that as well and let us know if this works for you and properly _fixes_ the problem. Thanks.

Yes, it worked for me.

In my package.json I updated `prisma` and `@prisma/client` with `4.12.0-integration-rtld-deepbind.3`

```json
  "dependencies": {
    "@prisma/client": "4.12.0-integration-rtld-deepbind.3",
    "prisma": "4.12.0-integration-rtld-deepbind.3"
  }
```

**Comment by aqrln** [comment]
This issue is fixed in **Prisma 4.13.0** which will be released tomorrow, on **Tuesday, Apr 18**. If you need to temporarily use a pre-release version until then, you can use `4.13.0-dev.33`.

---

For future readers:

Note that this GitHub issue is about a specific bug that occurs when using **Node.js 17+ on a system with OpenSSL 1.1 on x86-64 CPUs when using TLS to connect to database**. Please do not post comments here if you encounter a segfault under different circumstances, and **open a new issue instead**.

If you are facing segfaults on ARM, please leave a comment in https://github.com/prisma/prisma/issues/18510 or open a new issue.
